### PR TITLE
JWT Auth

### LINKED ISSUE #295: Create jwt authentication generator on Api apps
I'm going to turn my jwt implementation into a generator as discussed on gitter. This generator is intended for Api applications and it will need to do the following:

- Install [`crystal jwt`](https://github.com/crystal-community/jwt) shard
- Gen `SignUp::Create` action
- Gen `SignIn::Create` action
- Include `Authentic::PasswordAuthenticatable` in User model
- Gen `FindAuthenticable` form mixin
- Gen `PasswordValidations` form mixin
- Gen `GenerateToken` action mixin
- Gen `UserFromToken` action mixin
- Gen `AuthenticatedAction` abstract class

This assumes there is a `User` model so it should check for that and ask the user to generate the model and migrate the database.

There are also mixins created by `Authentic` for sending password resets that should also be generated for this, but maybe I'll tackle that as a separate issue.

### PR DESCRIPTION
**Pull request by paulcsmith**
- [x] Check for `token` body param
- [x] Properly check for `Bearer` token
- [x] Handle missing nested param `Missing nested params: user (Lucky::Exceptions::MissingNestedParam)`
- [x] Add note to `init` generator that says it includes JWT auth for API, remove browser text if API only
- [x] Add `/api/me` endpoint to show how to test with `RequireAuthToken`
- [x] `Errors::Show` should not render HTML page if it accepts json!

Closes https://github.com/luckyframework/lucky/issues/889
Closes #295

### CODE REVIEW

File: src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr [hunk 1]
```diff
@@ -0,0 +1,21 @@
+module Api::Auth::Helpers(T)
```
Comment (paulcsmith) [inline src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr]: Maybe don't pass in `T` and instead hardcode `User`? Probably more clear. This is what people will use 99% of the time and if they don't want it they can change it

File: src/api_authentication_app_skeleton/src/models/user_token.cr [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  ALGORITHM = JWT::Algorithm::HS256
+
+  def self.generate(user : User) : String
+    payload = {"user_id" => user.id}
+
+    settings.stubbed_token ||
+      JWT.encode(payload, Lucky::Server.settings.secret_key_base, ALGORITHM)
+  end
+
+  def self.decode(token : String)
```
Comment (paulcsmith) [inline src/api_authentication_app_skeleton/src/models/user_token.cr]: Maybe clarify this? `decode_user_id`?

File: src/generators/web.cr [hunk 3]
```diff
@@ -96,6 +95,10 @@ class LuckyCli::Generators::Web
     BaseAuthenticationSrcTemplate.new.render("./#{project_dir}", force: true)
   end
 
+  # TODO
+  private def add_api_authentication_to_src
+  end
```
Comment (paulcsmith) [inline src/generators/web.cr]: Need to actually generate this!

File: src/web_app_skeleton/src/serializers/errors/show_serializer.cr [hunk 4]
```diff
@@ -1,10 +1,14 @@
 # This is the default error serializer generated by Lucky.
 # Feel free to customize it in any way you like.
 class Errors::ShowSerializer < Lucky::Serializer
-  def initialize(@message : String, @details : String? = nil)
+  def initialize(
+    @message : String,
+    @details : String? = nil,
+    @param : String? = nil
```
Comment (paulcsmith) [inline src/web_app_skeleton/src/serializers/errors/show_serializer.cr]: This is what Stripe has in their error object and I think it is a good default. This way by default you can use `create!/update!` and Lucky will respond with the param that had the problem and what the error message is

File: src/web_app_skeleton/src/actions/errors/show.cr.ecr [hunk 5]
```diff
@@ -20,6 +20,20 @@ class Errors::Show < Lucky::ErrorAction
     <%- end -%>
   end
 
+  def handle_error(error : Avram::InvalidSaveOperationError)
+    if json?
```
Comment (paulcsmith) [inline src/web_app_skeleton/src/actions/errors/show.cr.ecr]: Only include this is for browser!

File: src/api_authentication_app_skeleton/spec/requests/api/me/show_spec.cr [hunk 6]
```diff
@@ -0,0 +1,17 @@
+require "../../../spec_helper"
+
+describe Api::Me::Show do
```
Comment (paulcsmith) [inline src/api_authentication_app_skeleton/spec/requests/api/me/show_spec.cr]: This is here to show how to test  endpoints that require auth

File: src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    end
+  end
+
+  private def auth_token : String?
+    bearer_token || token_param
+  end
+
+  private def bearer_token : String?
+    context.request.headers["Authorization"]?
+      .try(&.gsub("Bearer", ""))
```
Comment (paulcsmith) [inline src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr]: Works if using `Bearer` or not

File: src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+      .try(&.strip)
+  end
+
+  private def token_param : String?
+    params.get?(:auth_token)
```
Comment (paulcsmith) [inline src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr]: Also allows passing `auth_token` as  query param or body param

File: src/browser_app_skeleton/src/actions/browser_action.cr.ecr [hunk 9]
```diff
@@ -25,8 +25,8 @@ abstract class BrowserAction < Lucky::Action
   expose current_user
 
   # This method tells Authentic how to find the current user
-  private def find_current_user(id) : User
-    UserQuery.find(id)
+  private def find_current_user(id) : User?
+    UserQuery.new.id(id).first?
```
Comment (paulcsmith) [inline src/browser_app_skeleton/src/actions/browser_action.cr.ecr]: If the  user with that id was deleted, consider them signed out

File: src/web_app_skeleton/spec/setup/start_app_server.cr.ecr [hunk 10]
```diff
@@ -5,6 +5,8 @@ spawn do
 end
 
 at_exit do
+  <%- if browser? -%>
   LuckyFlow.shutdown
+  <%- end -%>
```
Comment (paulcsmith) [inline src/web_app_skeleton/spec/setup/start_app_server.cr.ecr]: This  file was moved to always be added whether you have  API mode or "full", but we only want LuckyFlow is using "full" mode

File: src/web_app_skeleton/spec/spec_helper.cr.ecr [hunk 11]
```diff
@@ -20,6 +20,7 @@ require "../db/migrations/**"
 require "./setup/**"
 
 include Carbon::Expectations
+include Lucky::RequestExpectations
```
Comment (paulcsmith) [inline src/web_app_skeleton/spec/spec_helper.cr.ecr]: Gives us `send_json` matcher

File: src/web_app_skeleton/src/actions/errors/show.cr.ecr [hunk 12]
```diff
@@ -20,6 +20,41 @@ class Errors::Show < Lucky::ErrorAction
     <%- end -%>
   end
 
+  def handle_error(error : Lucky::Exceptions::MissingNestedParam)
```
Comment (paulcsmith) [inline src/web_app_skeleton/src/actions/errors/show.cr.ecr]: Better JSON error message when param key is missing

File: src/web_app_skeleton/src/actions/errors/show.cr.ecr [hunk 13]
```diff
@@ -20,6 +20,41 @@ class Errors::Show < Lucky::ErrorAction
...
+    end
+    <%- end -%>
+  end
+
+  private def render_invalid_param_json(error)
+    param = error.save_operation_errors.keys.first.to_s
+    serialized_error = Errors::ShowSerializer.new(
+      "Invalid params",
+      param: param,
+      details: "#{param} #{error.save_operation_errors.values.first.first}"
```
Comment (paulcsmith) [inline src/web_app_skeleton/src/actions/errors/show.cr.ecr]: I'm gonna add some stuff to Avram  to make this clearer. Basically it is getting the first validation error for the first attribute with an error

File: src/web_app_skeleton/src/actions/errors/show.cr.ecr [hunk 14]
```diff
@@ -34,9 +69,7 @@ class Errors::Show < Lucky::ErrorAction
 
   # This is the catch all method that renders unhandled exceptions
   def handle_error(error : Exception) : Lucky::Response
-    Lucky.logger.error(unhandled_error: error.inspect_with_backtrace)
-
-    if Lucky::ErrorHandler.settings.show_debug_output
+    if Lucky::ErrorHandler.settings.show_debug_output && html?
```
Comment (paulcsmith) [inline src/web_app_skeleton/src/actions/errors/show.cr.ecr]: Only render debug page if HTML

**Comment by paulcsmith** [comment]
@watzon I think you mentioned wanting this. Feel free to take a look and LMK what you think

**Comment by watzon** [comment]
@paulcsmith tbh I don't remember 😁 but it is super exciting. Definitely something that will come in handy.

**Comment by paulcsmith** [comment]
Are you sure nothing is glaring at you...

glaring

**Comment by jwoertink** [comment]
FINKLE IS EINHORN!!!

File: src/browser_app_skeleton/src/actions/browser_action.cr.ecr (hunk 9 above)
Comment (paulcsmith) [reply src/browser_app_skeleton/src/actions/browser_action.cr.ecr]: https://github.com/luckyframework/lucky/issues/889

File: src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr (hunk 7 above)
Comment (jwoertink) [reply src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr]: We just need to make sure we add guides that show how to update this. (i.e. you don't want param tokens, and require the word "Bearer" to be present in the header, etc...)

**Comment by jwoertink** [comment]
oh, and one last thing missing is a response when the token is missing or bad. There's a nice json response when you forget to add the Accept header

```
[14:07PM] throw_this_away (master)$ curl -H "Accept: application/json" -H "Authorization: wrong" localhost:5000/api/me
[14:07PM] throw_this_away (master)$ curl -H "Accept: application/json" localhost:5000/api/me
[14:09PM] throw_this_away (master)$ curl -H "Accept: application/json" localhost:5000/api/me?auth_token=wrong
[14:09PM] throw_this_away (master)$ curl -H "Accept: application/json" localhost:5000/api/me?auth_token=right
{"email":"test@test.com"}
[14:09PM] throw_this_away (master)$ curl -H "Accept: application/json" localhost:5000
[14:09PM] throw_this_away (master)$ curl localhost:5000/api/me?auth_token=right
{"message":"An unexpected error occurred","param":null,"details":null}
```

File: src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr (hunk 7 above)
Comment (paulcsmith) [reply src/api_authentication_app_skeleton/src/actions/mixins/api/auth/helpers.cr]: Yeah great idea

File: src/lucky_cli/wizard/web.cr [hunk 15]
```diff
@@ -66,17 +67,30 @@ class LuckyCli::Wizard::Web
...
-    Lucky can be generated with email and password authentication
+      Lucky can be generated with email and password authentication with a token
 
-     ● Sign in and sign up
-     ● Mixins for requiring sign in
-     ● Password reset
-     ● Generated files can easily be removed/customized later
+      ● Sign in and sign up endpoints that return a JWT with the user id
+      ● Mixins for requiring an auth token for endpoints
+      ● Generated files can easily be removed/customized later
```
[the remark this answers is not in the document]
Comment (paulcsmith) [reply src/lucky_cli/wizard/web.cr]: Nice catch!

**Comment by paulcsmith** [comment]
When you were seeing the `Accept` header error, did you generate an app with `full` or with `API`. I think with `API` it should default to  application/json but  I can't remember

File: src/api_authentication_app_skeleton/src/actions/mixins/api/auth/require_auth_token.cr [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    )
+  end
+
+  private def auth_error_details : String
+    if auth_token
+      "The provided authentication token was incorrect."
+    else
+      "An authentication token is required. Please include a token in an 'auth_token' param or 'Authorization' header."
+    end
+  end
```
Comment (paulcsmith) [inline src/api_authentication_app_skeleton/src/actions/mixins/api/auth/require_auth_token.cr]: Added more details to the auth message!

**Pull request merged**